### PR TITLE
Refactor `[@@unboxed]` patch 1/2: move a boolean

### PR DESCRIPTION
**Pull request by stedolan**
Type declarations include a boolean `type_unboxed.unboxed` as well as a representation `type_kind`. The two are not orthogonal:

  - If `type_kind = Type_abstract`, `type_unboxed.unboxed` must be `false`
  - If `type_kind = Type_open`, `type_unboxed.unboxed` must be `false`
  - If `type_kind = Type_record (lbls, rep)`, `type_unboxed.unboxed` must be true iff `rep` is `Record_unboxed _`
  - If `type_kind = Type_variant cstrs`, then `type_unboxed.unboxed` indicates whether this variant is unboxed

This patch moves this boolean into `Type_variant`, adding a `variant_representation` field to match the existing `record_representation` field, indicating whether the variant is unboxed or not.

The only functional change is that some checks are done in a different order, so if a program is broken in multiple ways the exact one that the compiler chooses to complain about may now differ. For instance, this program used to complain about the unboxed annotations not matching, and now it complains about the field types not matching:
```ocaml
module M : sig
  type t = { foo : int }
end = struct
  type t = { foo : string } [@@unboxed]
end
```

### CODE REVIEW

File: typing/ctype.ml [hunk 1]
```diff
@@ -2371,7 +2371,7 @@ and mcomp_type_decl type_pairs env p1 p2 tl1 tl2 =
       | Type_record (lst,r), Type_record (lst',r') when r = r' ->
           mcomp_list type_pairs env tl1 tl2;
           mcomp_record_description type_pairs env lst lst'
-      | Type_variant v1, Type_variant v2 ->
+      | Type_variant (v1,r), Type_variant (v2,r') when r = r' ->
```
Comment (gasche) [inline typing/ctype.ml:2390]: This change looks like it may catch a bug. Does it?

File: typing/includecore.ml [hunk 2]
```diff
@@ -429,16 +426,22 @@ let type_declarations ?(equality = false) ~loc env ~mark name
...
+        | Variant_unboxed, Variant_regular ->
+           Some (Unboxed_representation First)
+        | Variant_regular, Variant_unboxed ->
+           Some (Unboxed_representation Second)
+        | _ ->
+          Option.map
+            (fun var_err -> Variant_mismatch var_err)
+            (compare_variants ~loc env decl1.type_params decl2.type_params 1
+               cstrs1 cstrs2)
+        end
```
Comment (gasche) [inline typing/includecore.ml]: I think that we want a `compare_variants_with_representation` help here, by symmetry with the record case.

File: typing/datarepr.ml [hunk 3]
```diff
@@ -126,7 +121,7 @@ let constructor_descrs ~current_unit ty_path decl cstrs =
         let cstr_name = Ident.name cd_id in
         let existentials, cstr_args, cstr_inlined =
           let representation =
-            if decl.type_unboxed.unboxed
+            if rep = Variant_unboxed
             then Record_unboxed true
             else Record_inlined idx_nonconst
```
Comment (gasche) [inline typing/datarepr.ml]: I would prefer if this code matched on `rep`.

File: typing/datarepr.ml [hunk 4]
```diff
@@ -116,7 +111,7 @@ let constructor_descrs ~current_unit ty_path decl cstrs =
         in
         let (tag, descr_rem) =
           match cd_args with
-          | _ when decl.type_unboxed.unboxed ->
+          | _ when rep = Variant_unboxed ->
```
Comment (gasche) [inline typing/datarepr.ml]: I would prefer `match rep, cd_args`, which would avoid the implicit fragile pattern matching here.

File: typing/includecore.ml [hunk 5]
```diff
@@ -374,9 +374,15 @@ let compare_records_with_representation ~loc env params1 params2 n
...
-  | err -> err
+      Some (match rep1, rep2 with
+            | _, Record_unboxed _ -> Unboxed_representation Second
+            | Record_unboxed _, _ -> Unboxed_representation First
+            | _, Record_float ->
+               Record_mismatch (Unboxed_float_representation Second)
+            | _, _ ->
+               Record_mismatch (Unboxed_float_representation First))
+  | None -> None
+  | Some err -> Some (Record_mismatch err)
```
Comment (gasche) [inline typing/includecore.ml:410]: Idle note: this need not be part of this PR, but this code would be nicer with `(unit, ...) result` instead of `... option`.

File: typing/includecore.ml [hunk 6]
```diff
@@ -374,9 +374,15 @@ let compare_records_with_representation ~loc env params1 params2 n
...
-      let pos = if rep2 = Record_float then Second else First in
-      Some (Unboxed_float_representation pos)
-  | err -> err
+      Some (match rep1, rep2 with
+            | _, Record_unboxed _ -> Unboxed_representation Second
+            | Record_unboxed _, _ -> Unboxed_representation First
+            | _, Record_float ->
+               Record_mismatch (Unboxed_float_representation Second)
+            | _, _ ->
+               Record_mismatch (Unboxed_float_representation First))
```
Comment (gasche) [inline typing/includecore.ml]: If I understand correctly, this change fixes a bug in the `Regular, Unboxed` case for example. Is this correct?

The code (old or new) is making assumptions about the fact that the `Record_inlined` and `Record_extension` constructors will not show up here. Are those assumptions correct? Are they documented somewhere?

I wonder if we could write a non-fragile pattern-match instead of first check `rep1 <> rep2` and then coming up with a pithy list of relevant cases.

File: typing/includecore.ml [hunk 7]
```diff
@@ -429,16 +426,22 @@ let type_declarations ?(equality = false) ~loc env ~mark name
...
-        Option.map
-          (fun var_err -> Variant_mismatch var_err)
-          (compare_variants ~loc env decl1.type_params decl2.type_params 1
-             cstrs1 cstrs2)
+        begin match rep1, rep2 with
+        | Variant_unboxed, Variant_regular ->
+           Some (Unboxed_representation First)
+        | Variant_regular, Variant_unboxed ->
+           Some (Unboxed_representation Second)
+        | _ ->
```
Comment (gasche) [inline typing/includecore.ml]: Could we use a non-fragile match here?

File: typing/printtyp.ml [hunk 8]
```diff
@@ -1277,30 +1277,33 @@ let rec tree_of_type_decl id decl =
...
-            tree_of_typexp false ty, decl.type_private
+            tree_of_typexp false ty, decl.type_private, false
         end
-    | Type_variant cstrs ->
+    | Type_variant (cstrs, rep) ->
         tree_of_manifest (Otyp_sum (List.map tree_of_constructor cstrs)),
-        decl.type_private
-    | Type_record(lbls, _rep) ->
+        decl.type_private,
+        (rep = Variant_unboxed)
```
Comment (gasche) [inline typing/printtyp.ml:1313]: I would prefer a pattern-matching as below. (I would even prefer a non-fragile pattern, but here I think it's a matter of taste and there are reasonable arguments for the fragile version.)

File: typing/typedecl_unboxed.ml [hunk 9]
```diff
@@ -34,20 +34,16 @@ let rec get_unboxed_type_representation env ty fuel =
...
+       | Type_variant ([{cd_args = Cstr_tuple [ty2]; _}], Variant_unboxed)
+       | Type_variant ([{cd_args = Cstr_record [{ld_type = ty2; _}]; _}],
+                       Variant_unboxed)}
       ->
         let ty2 = match ty2.desc with Tpoly (t, _) -> t | _ -> ty2 in
         get_unboxed_type_representation env
           (Ctype.apply env type_params ty2 args) (fuel - 1)
-    | {type_kind=Type_abstract} -> Unavailable
-          (* This case can occur when checking a recursive unboxed type
-             declaration. *)
```
Comment (gasche) [inline typing/typedecl_unboxed.ml:49]: Why did this case disappear? I suppose it refers to situations where we were not about to locate/expand the definition of a type abbreviation that is marked as unboxed?

File: typing/types.mli [hunk 10]
```diff
@@ -370,14 +370,15 @@ type type_declaration =
     type_loc: Location.t;
     type_attributes: Parsetree.attributes;
     type_immediate: Type_immediacy.t;
-    type_unboxed: unboxed_status;
+    type_unboxed_default: bool;
+    (* true if the unboxed-ness of this type was chosen by a compiler flag *)
```
Comment (gasche) [inline typing/types.mli:374]: Two remarks here:
- do you remember what this information is used for?
- another (nicer?) approach would be to store this information as a parameter of `Record_unboxed` and `Variant_unboxed`, for example

```ocaml
type unboxed_setting = On_decl | Global_flag
```

File: typing/ctype.ml (hunk 1 above)
Comment (stedolan) [reply typing/ctype.ml:2390]: It is a mild improvement of behaviour, but I would not go so far as to call the old behaviour buggy.

`mcomp` is used during GADT matches to determine that two types cannot possibly be equal (even via internal definitions not currently visible). It can already determine that two variant types with different constructor names or arities cannot be equal, and now it can also determine that two variant types that differ only in unboxiness cannot be equal.

This makes it marginally better at refuting impossible cases, but the old behaviour was also sound.

File: typing/includecore.ml (hunk 2 above)
Comment (stedolan) [reply typing/includecore.ml]: Done. I ended up rewriting both of these functions.

File: typing/datarepr.ml (hunk 3 above)
Comment (stedolan) [reply typing/datarepr.ml]: Good point. (I was making fairly mechanical replacements here, didn't notice that this should really be a match)

File: typing/datarepr.ml (hunk 4 above)
Comment (stedolan) [reply typing/datarepr.ml]: Done

File: typing/includecore.ml (hunk 6 above)
Comment (stedolan) [reply typing/includecore.ml]: I wrote the pattern match, it is nicer. There's still an assumption that Record_inlined and Record_extension don't mix, but now it's an explicit assert false rather than silently converting errors to weird messages about floats.

(For the record, I believe the assumption here is true because it's impossible to name an inlined record in the source language, and the compiler keeps the autogenerated inline record types separate)

File: typing/includecore.ml (hunk 7 above)
Comment (stedolan) [reply typing/includecore.ml]: Done, see above

File: typing/printtyp.ml (hunk 8 above)
Comment (stedolan) [reply typing/printtyp.ml:1313]: Changing fragile equalities to non-fragile pattern matches I can get behind, but what's the advantage of a fragile pattern match over an equality check?

File: typing/typedecl_unboxed.ml (hunk 9 above)
Comment (stedolan) [reply typing/typedecl_unboxed.ml:49]: The case this code handles (a type which, while abstract, is somehow known to be unboxed) is no longer expressible. Abstract types generated while making recursive unboxed types no longer have any special tagging.

In these cases, the return value of this function changes from `Unavailable` to `This t`, where `t` is an abstract type. The only uses of the return value of this function are to compute its immediacy or compare with a builtin type, and in both of those cases `Unavailable` and `This <abstract>` are not distinguished.

(I think the interface to this function is poor, and the second patch in this series will change it)

File: typing/printtyp.ml (hunk 8 above)
Comment (gasche) [reply typing/printtyp.ml:1313]: The record case below has very similar code but uses a fragile pattern. I think the code would be more consistent if they both used the same approach (and records cannot use equality directly because it takes an argument), but this is nitpicking territory and you've already been very kind to add many robust matchings, so I'll stay quiet on this one. Thanks!

File: typing/types.mli (hunk 10 above)
Comment (stedolan) [reply typing/types.mli:374]: It is used in exactly one place, in the implementation of Warning 61:
```
# type t = Foo of int;;
type t = Foo of int
# external blah : int -> t = "blah";;
Warning 61: This primitive declaration uses type t, whose representation
may be either boxed or unboxed. Without an annotation to indicate
which representation is intended, the boxed representation has been
selected by default. This default choice may change in future
versions of the compiler, breaking the primitive implementation.
You should explicitly annotate the declaration of t
with [@@boxed] or [@@unboxed], so that its external interface
remains stable in the future.
external blah : int -> t = "blah"
```
It's a sufficiently obscure and special-case situation that I'd rather not clutter up the frequently used representation information with it.

File: typing/datarepr.ml (hunk 4 above)
Comment (gasche) [reply typing/datarepr.ml]: (I still see `when rep = ..`, did you forgot to update your remote branch?)

**Comment by stedolan** [comment]
Pushed a version with fixes, build should hopefully pass this time. (I generally build locally without ocamldoc as it's much faster, but there was a single match on `Type_variant` in ocamldoc that broke)

There's a tiny functional change in this version, which is visible in the testsuite: if a variant definition fails to match its signature by differing both in type and unboxiness, the compiler will now complain about the type mismatch rather than the `[@@unboxed]` mismatch. (This fell out from the way the rewritten `compare_variants` code worked, but I prefer the new behaviour anyway)

**Comment by gasche** [comment]
> The compiler will now complain about the type mismatch rather than the [@@unboxed] mismatch. (This fell out from the way the rewritten compare_variants code worked, but I prefer the new behaviour anyway).

Yes, that sounds better.

File: typing/includecore.ml [hunk 11]
```diff
@@ -372,11 +387,28 @@ and compare_records ~loc env params1 params2 n
...
+  | None, Record_regular, Record_regular
+  | None, Record_float, Record_float
+  | None, Record_unboxed _, Record_unboxed _
+  | None, Record_inlined _, Record_inlined _
+  | None, Record_extension _, Record_extension _ ->
+     None
+  | Some err, _, _ ->
+     Some (Record_mismatch err)
+  | None, Record_unboxed _, _ ->
+     Some (Unboxed_representation First)
```
Comment (gasche) [inline typing/includecore.ml]: I apologies for the nitpick (plus side: I'm looking at this PR again!), but I find this pattern-match hard to read. At this point you know that the `_` in the third component is a representation distinct from `Record_unboxed`, but this is not so clear in the code. Here would be a concrete proposal (that could also be applied in `compare_variants`, even though it is shorter so more readable):

- deal with the `Some err` case first
- then group all the cases that deal with `Record_unboxed` together (success or failure), and then all the cases about `Record_float` together. (You can use an empty line between groups of cases.)

File: typing/includecore.ml [hunk 12]
```diff
@@ -372,11 +387,28 @@ and compare_records ~loc env params1 params2 n
...
+  | None, _, Record_unboxed _ ->
+     Some (Unboxed_representation Second)
+  | None, Record_float, _ ->
+     Some (Record_mismatch (Unboxed_float_representation First))
+  | None, _, Record_float ->
+     Some (Record_mismatch (Unboxed_float_representation Second))
+  | None,
+    (Record_regular|Record_inlined _|Record_extension _),
+    (Record_regular|Record_inlined _|Record_extension _) ->
+     assert false
```
Comment (gasche) [inline typing/includecore.ml]: a comment on why we `assert false` here could be nice

File: typing/typedecl.ml [hunk 13]
```diff
@@ -78,10 +78,10 @@ let get_unboxed_from_attributes sdecl =
...
   match boxed, unboxed, !Clflags.unboxed_types with
   | true, true, _ -> raise (Error(sdecl.ptype_loc, Boxed_and_unboxed))
-  | true, false, _ -> unboxed_false_default_false
-  | false, true, _ -> unboxed_true_default_false
-  | false, false, false -> unboxed_false_default_true
-  | false, false, true -> unboxed_true_default_true
+  | true, false, _ -> false, false
+  | false, true, _ -> true, false
+  | false, false, false -> false, true
+  | false, false, true -> true, true
```
Comment (gasche) [inline typing/typedecl.ml]: It's unclear what this function does, a comment could help (or a record type just for this function?). I also believe that the comments lines 73-74 is outdated and should be removed.

**Review by gasche — APPROVED** [review]
This is good to go, possibly with changes suggested in comments.

**Comment by gasche** [comment]
@stedolan gentle ping. This is not super important, but we risk adding more work if we wait too long.

File: typing/includecore.ml (hunk 11 above)
Comment (stedolan) [reply typing/includecore.ml]: I think this did make it clearer. Splitting it into two separate matches also helped.

File: typing/includecore.ml (hunk 12 above)
Comment (stedolan) [reply typing/includecore.ml]: I have no idea why we assert false here! In trunk at the moment, if you can manage to get this code to compare, say, a `Record_regular` and a `Record_inlined ...`, then it will error with a nonsensical message about unboxed floats. If this situation is possible, then an assertion failure is an improvement.

File: typing/typedecl.ml (hunk 13 above)
Comment (stedolan) [reply typing/typedecl.ml]: On closer inspection, the type was just wrong. This function now returns a bool option, not a bool * bool.

**Comment by gasche** [comment]
Thanks for taking the review comments into account. I am eager to see this general arc integrated, so I went ahead and merged. (Some commits would have been squashable, but not the bootstrap. Oh well.)
